Add vertical scale answer format

rsanchezsaez opened the pull request:
I have implemented vertical scale support ([issue #3](https://github.com/ResearchKit/ResearchKit/issues/3)).

`ORKScaleSlider` now has the `isVertical` property. When set to `YES`, it internally uses the `UIView`'s`transform` property and some minor *bounds* and *touch* trickery to draw and work in vertical.

`ORKScaleSliderView` now has vertical support through the use of the new `isVertical` method from the `ORKScaleAnswerFormatProvider`.

New classes: `ORKVerticalScaleAnswerFormat` and `ORKContinuousVerticalScaleAnswerFormat`. The corresponding example steps have been added to `ORKCatalog`'s *Scale Question*.

rsanchezsaez commented:
I tried to visually layout vertical scales so the scrolling is minimum on *iPhone 4S* sizes, but this produces somewhat short vertical scales. Maybe we should layout to *iPhone 5* sizes and just have *iPhone 4S* users scrolling a bit.

Also, I'm not happy by the fact I had to increase the `ORKSurveyAnswerCellForScale`'s `suggestedCellHeightForView:` (shared by both vertical and horizontal scales) so it fits the vertical ones.

Ideally, the height wouldn't increase for the horizontal scale cell (and vertical ones could be made even taller). But I couldn't figure a simple and sound way of doing this without a major refactor (since the `ORKSurveyAnswerCellForScale` class is inferred from `ORKQuestionTypeScale`, and it doesn't make much sense to add an additional `ORKQuestionTypeVerticalScale` value just for the vertical scales). Any suggestion regarding this issue?

3lvis commented on ResearchKit/Common/ORKAnswerFormat.m:
```diff
@@ -251,11 +251,16 @@ @implementation ORKAnswerFormat
                                                   step:(NSInteger)step
-                                          defaultValue:(NSInteger)defaultValue
-{
+                                          defaultValue:(NSInteger)defaultValue {
```
Better be consistent with the location of the `{` (put it in the next line)

3lvis commented on ResearchKit/Common/ORKAnswerFormat.m:
```diff
@@ -251,11 +251,16 @@ @implementation ORKAnswerFormat
++ (ORKScaleAnswerFormat *)verticalScaleAnswerFormatWithMaxValue:(NSInteger)scaleMax
+                                                       minValue:(NSInteger)scaleMin
+                                                           step:(NSInteger)step
+                                                   defaultValue:(NSInteger)defaultValue {
```
Better be consistent with the location of the `{` (put it in the next line)

YuanZhu-apple commented on ResearchKit/Common/ORKAnswerFormat.h:
```diff
@@ -250,6 +259,21 @@ ORK_CLASS_AVAILABLE
+ORK_CLASS_AVAILABLE
+@interface ORKVerticalScaleAnswerFormat : ORKScaleAnswerFormat
+
+@end
```
Would it be much simpler to just add a `vertical` boolean property on  `ORKScaleAnswerFormat` and `ORKContinuousScaleAnswerFormat `?

YuanZhu-apple commented on ResearchKit/Common/ORKSurveyAnswerCellForScale.m:
```diff
@@ -119,9 +119,7 @@ - (IBAction)sliderValueChanged:(id)sender {
-    return 140.0;
+    return 244.0;
 }
 
```
I think better to add a second `vertical` parameter to this method, so it can return different height for different layout.

jwe-apple replied:
Yuan - not sure quite how that would work! This is a class method on the base class, and vertical/horizontal is very specific to this class. It might be easiest to just create a subclass of ORKSurveyAnswerCellForScale to cover the vertical case.

YuanZhu-apple commented:
Congratulations on the FIRST new feature PR for ResearchKit!

To the visually layout of vertical scale, scale bar should has an adequate height to assist user to select on the bar easier.
* Value labels should be move to the left to save space for scale bar
* It's OK to have a fixed height across different screen sizes, but it is better to make the page is not scrollable on iPhone 6 and iPhone 6 plus (page scrolling on iphone 4s and iphone 5 is fine). Try this magic cell height: `358.0`.

I did an engineering mockup:
![mockup](https://cloud.githubusercontent.com/assets/11466704/7169824/de1e29ce-e37d-11e4-8277-c8ccf2d6b81c.png)

jwe-apple commented:
One more comment - to accept this pull request, it will also need to work properly with VoiceOver enabled. The existing horizontal control has VoiceOver support, so this should be straightforward.

rsanchezsaez commented:
Hi guys,

thanks for the feedback. I'll look into the design changes. Definitely the end labels on the side of the slider is a good idea that saves some vertical space.

I tried displaying `_valueLabel` to the left of the vertical scale before but it didn't look quite right (all the rest of the answer formats have a nice vertical-axis hierarchical organisation to their UI elements, and I thought this setup kind of broke it). I'll try again and report back.

Regarding the height of `ORKSurveyAnswerCellForScale`, I had the problem that @jwe-apple mentions: it's quite tricky to send parameters to the `suggestedCellHeightForView:` method without a big refactoring. Creating a `ORKSurveyAnswerCellForVerticalScale` subclass may be the best solution. Note that I'll have to add a (otherwise needless) `ORKQuestionTypeVerticalScale` value for this to work.

Regarding adding a `isVertical` property to `ORKScaleAnswerFormat` and ` ORKContinuousScaleAnswerFormat` instead of adding the `ORKVerticalScaleAnswerFormat` and `ORKContinuousVerticalScaleAnswerFormat` classes, I thought the the additional classes in my solution were neater from the framework user perspective (and would allow for more encapsulated visual customization in the subclasses). But tell me if the adding property makes more sense and I'll modify that, this should be pretty straight forward.

jwe-apple commented:
We've found the answer formats tend to be pretty static and think a property makes sense. Probably read-only, but could be convinced otherwise.

rsanchezsaez commented:
A read-only property passed on the answer format initializers makes sense. I'll make the changes soon.

Do you prefer `isVertical` or `vertical` for the property names (for `ORKScaleSlider`, `ORKScaleAnswerFormatProvider`, `ORKScaleAnswerFormat`, and `ORKContinuousScaleAnswerFormat`)? Looking at past usages in other Apple code: *iOS*' `UITextSelectionRect` uses `isVertical`, but *OS X*'s `NSSplitView` uses `vertical`.

[Coding Guidelines for Cocoa](https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingIvarsAndTypes.html#//apple_ref/doc/uid/20001284-BAJGIIJE) naming conventions seem to recommend using the `vertical` name with a custom specified `isVertical` get accessor as in:

    @property (assign, getter=isEditable) BOOL editable;

jwe-apple commented:
Yep, the latter

rsanchezsaez commented:
I have added commits implementing the requested feedback.

Regarding visual layout, I optimized it for iPhone 6 and up. I moved the range labels to the left side of the scale. I left the `valueLabel` at the top center-y of the view, as I feel it looks nicer this way. I think that with the default iPhone 6 cell height, there's enough room for the vertical scale. If you disagree, let me know and I can move the label to center-x left-y of the view.

I think that VoiceOver should work straight away as well (I checked with the Accessibility inspector). I cannot test on an actual device right now, can anybody confirm that it works?

Thanks.

rsanchezsaez commented:
Here are some iPhone 6 sized screenshots:

![vertical scale #1](https://cloud.githubusercontent.com/assets/444313/7207016/31e8495c-e52e-11e4-86e0-0b960618b589.png)

---

![vertical scale #2](https://cloud.githubusercontent.com/assets/444313/7207018/320f453e-e52e-11e4-8c72-580b0906eca7.png)

---

![continuous vertical scale](https://cloud.githubusercontent.com/assets/444313/7207017/32089e00-e52e-11e4-86c4-4db84e1b9ecd.png)

3lvis commented:
Kicking ass @rsanchezsaez!

:clap:

avocade commented:
This is such a great read! And even greater to have Apple engineers on Github—never thought I'd see the day, but I'm glad I was wrong to doubt.

YuanZhu-apple commented:
One more thing, `CurrentThumbImage` should be rotated to let its drop shadow point downward.

rsanchezsaez commented:
Excellent catch! Issue solved.

YuanZhu-apple commented on ResearchKit/Common/ORKAnswerFormat.h:
```diff
@@ -136,15 +139,17 @@ ORK_CLASS_AVAILABLE
+                                                                   minimumValue:(double)scaleMinimum
+                                                                   defaultValue:(double)defaultValue
+                                                          maximumFractionDigits:(NSInteger)maximumFractionDigits
+                                                                       vertical:(BOOL)verticalFlag;
```
Personally, I prefer `vertical` over `verticalFlag`.

YuanZhu-apple commented:
`ORKTest App`(under `Testing` directory) also needs to be updated. Nice to add few vertical scale testing steps in `scale` task.

![ios simulator screen shot apr 20 2015 12 41 58 pm](https://cloud.githubusercontent.com/assets/11466704/7238920/dc2b3464-e75a-11e4-8ec9-55220ebeb1b6.png)

rsanchezsaez commented:
Added two vertical scale tests to the `scale` task in `ORKTest`.

Also added `vertical` property to `NSCoding`, and added it to the serialization tests. Finally, renamed some remaining `verticalFlag` arguments to `vertical`.

jwe-apple commented:
Thanks, will review shortly.

jwe-apple commented on ResearchKit/Common/ORKQuestionStepViewController.m:
```diff
@@ -784,8 +785,11 @@ - (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPa
+        case ORKQuestionTypeVerticalScale:{
+            height = [ORKSurveyAnswerCellForVerticalScale suggestedCellHeightForView:tableView];
+        }
+            break;
```
Rather than adding this question type, I'd prefer to change the logic for ORKQuestionTypeScale to interrogate the impliedAnswerFormat of the question step, cast appropriately, and check the vertical property. It might be best to add a method to questionStep like isFormatVertical, which makes the appropriate check. This would be in line with the above code for ORKQuestionTypeSingle/MultipleChoice.

rsanchezsaez replied:
Yeah, I never felt good about adding a new `ORKQuestionTypeVerticalScale` value. I didn't realize `ORKQuestionStep` had this helper methods. I will improve it.

jwe-apple commented on ResearchKit/Common/ORKAnswerFormat.h:
```diff
@@ -45,6 +45,9 @@ typedef NS_ENUM(NSInteger, ORKQuestionType) {
     ORKQuestionTypeScale,
     
+    /// The vertical scale question type asks participants to place a mark at an appropriate position on a continuous or discrete vertical line.
+    ORKQuestionTypeVerticalScale,
```
Would prefer not to add this as a separate question type. See further comment below.

jwe-apple commented on ResearchKit/Common/ORKQuestionStepViewController.m:
```diff
@@ -563,6 +563,7 @@ - (ORKSurveyAnswerCell *)answerCellForTableView:(UITableView *)tableView
     static dispatch_once_t onceToken;
     dispatch_once(&onceToken, ^{
         typeAndCellMapping = @{@(ORKQuestionTypeScale): [ORKSurveyAnswerCellForScale class],
+                               @(ORKQuestionTypeVerticalScale): [ORKSurveyAnswerCellForVerticalScale class],
```
Instead of doing this, you can do a special case correction in the code below this to handle the vertical case.

jwe-apple commented on ResearchKit/Common/ORKScaleSlider.m:
```diff
@@ -96,10 +164,9 @@ - (void)updateValueForTouchAtPoint:(CGPoint)touchPoint {
-    CGFloat centerY = [self bounds].size.height / 2.0;
+- (void)drawRect:(CGRect)rect {
+    CGRect trackRect = [self trackRectForBounds:self.bounds];
+    CGFloat centerY = self.bounds.size.height / 2.0;
```
This is ok, but we do prefer the bounds property calls to be written as method calls to make the fact that they're method calls explicit. If I were going to make a change here, for instance, I'd factor out the call to the bounds method and then re-use the struct for both the following calculations.

rsanchezsaez replied:
Ah, I see. I replaced it to a *property getter call* because of `UIView` public documentation defining it as a *property*. I'll change it back.

jwe-apple replied:
It is a property, but ObjC property accesses are still method calls, and they're still relatively expensive compared to accessing (for example) a member of a struct.

rsanchezsaez replied:
I see, thanks for the clarification. :)

jwe-apple commented on ResearchKit/Common/ORKScaleSlider.m:
```diff
@@ -60,17 +63,78 @@ - (id)initWithFrame:(CGRect)frame
+    CGRect thumbRect = [self thumbRectForBounds:self.bounds trackRect:trackRect value:self.value];
+    for (UIView *subview in self.subviews) {
+        if (CGRectEqualToRect(thumbRect, subview.frame)) {
+            thumbImageSubview = subview;
```
This is ugly and error-prone. It would be better to have and explicitly set a thumb image asset. I'm willing to accept in the short term, though. Please add a comment to this effect and open an issue for fixing in future.

rsanchezsaez replied:
I agree that it's hackish and not very future proof, but in my opinion the usage of the actual `thumbRect` for identifying the subview makes it quite safe. Doesn't it?

Anyway, yeah, using an explicit thumb would be better, but I didn't wan't to bother you guys for a thumb design. I'll add a comment.

jwe-apple commented on ResearchKit/Common/ORKScaleSlider.m:
```diff
@@ -60,17 +63,78 @@ - (id)initWithFrame:(CGRect)frame
+    UIView *view = nil;
+    if (_vertical) {
+        // In vertical mode, we need to ignore the touch area for the needed extra width
+        const CGFloat desiredSliderWidth = 36.0;
```
Can we widen this a little, perhaps to 44?

rsanchezsaez commented:
All feedback issues taken care of.

jwe-apple commented:
With that one naming change I'll merge this.

rsanchezsaez commented:
Done.

jwe-apple commented:
Thanks, great job on this!

rsanchezsaez commented:
Thanks for the thoughtful comments all along. :+1:

rsanchezsaez commented:
Hey, I think I made a mistake in `ORKQuestionStep`. One needs to do:

    return (ORKDynamicCast([self impliedAnswerFormat], ORKScaleAnswerFormat).vertical ||
            ORKDynamicCast([self impliedAnswerFormat], ORKContinuousScaleAnswerFormat).vertical);

(the property name) instead of:

    return (ORKDynamicCast([self impliedAnswerFormat], ORKScaleAnswerFormat).isVertical ||
            ORKDynamicCast([self impliedAnswerFormat], ORKContinuousScaleAnswerFormat).isVertical);

(the getter name).

If you clean the project and recompile, *Xcode* is not happy. I've fixed it in the latest commit in [the corresponding branch](https://github.com/rsanchezsaez/ResearchKit/tree/rsanchezsaez-verticalscale). I think you can just do a fast-forward merge to keep *GitHub* clean, but tell me if you want me to submit a separate pull request.

I spotted it when trying to compile the *merged master* branch. I'm not sure why *Xcode* was happily compiling on my separate branch.

rsanchezsaez commented:
Oh, I figured why it doesn't compile now. [Pull request #48](https://github.com/ResearchKit/ResearchKit/pull/48) (Fix double evaluation in ORKDynamicCast macro) makes the compiler to see the general `id` type, rather than the casted type, coming from the  `ORKDynamicCast()` macro. Not sure if [PR#48](https://github.com/ResearchKit/ResearchKit/pull/48) needs to be re-evaluated or not.

Still, the fix to my code still applies, I think.

rsanchezsaez commented:
Actually, my previous fix does not compiler either. Explicitly calling the getter works:

    return ([ORKDynamicCast([self impliedAnswerFormat], ORKScaleAnswerFormat) isVertical] ||
            [ORKDynamicCast([self impliedAnswerFormat], ORKContinuousScaleAnswerFormat) isVertical]);

I've overwritten my fixing commit to reflect this. Although I think we may want to go back to an `ORKDynamicCast()` implementation that returns a properly casted object (and then I'd prefer to just use the `.vertical` property syntax).

Sorry about the confusion.

jwe-apple commented:
Don't panic. I think the ORKDynamicCast() just needs extra parens. I'll fix it up.

jwe-apple commented:
Just pushed my fix to master, thanks!

rsanchezsaez commented:
Perfect.

drekryan commented:
@jwe-apple :+1: Thanks for that fix!
